etcdserver/api/v3rpc/watch.go

```diff
@@ -108,69 +111,88 @@ func (ws *watchServer) Watch(stream pb.Watch_WatchServer) error {
 	defer sws.close()
 
 	go sws.sendLoop()
-	return sws.recvLoop()
-}
-
-func (sws *serverWatchStream) recvLoop() error {
-	for {
-		req, err := sws.gRPCStream.Recv()
-		if err == io.EOF {
-			return nil
-		}
-		if err != nil {
-			return err
+	errc := sws.recvLoop()
+	select {
+	case err := <-errc:
+		return err
+	case <-stream.Context().Done():
+		err := stream.Context().Err()
+		// the only server-side cancellation is noleader for now.
+		if err == context.Canceled {
+			return rpctypes.ErrGRPCNoLeader
 		}
+		return err
+	}
+}
 
-		switch uv := req.RequestUnion.(type) {
-		case *pb.WatchRequest_CreateRequest:
-			if uv.CreateRequest == nil {
-				break
-			}
+func (sws *serverWatchStream) recvLoop() chan error {
+	errc := make(chan error, 1)
 
-			creq := uv.CreateRequest
-			if len(creq.Key) == 0 {
-				// \x00 is the smallest key
-				creq.Key = []byte{0}
-			}
-			if len(creq.RangeEnd) == 1 && creq.RangeEnd[0] == 0 {
-				// support  >= key queries
-				creq.RangeEnd = []byte{}
-			}
-			wsrev := sws.watchStream.Rev()
-			rev := creq.StartRevision
-			if rev == 0 {
-				rev = wsrev + 1
-			}
-			id := sws.watchStream.Watch(creq.Key, creq.RangeEnd, rev)
-			if id != -1 && creq.ProgressNotify {
-				sws.progress[id] = true
+	go func() {
```

Comment:
can `recvLoop` stay the same and instead move the errc/goroutine machinery into `Watch`?